packages/server/postgres/utils/checkEqBase.ts

```diff
@@ -0,0 +1,125 @@
+  [Property in keyof Partial<rethinkType>]: [CustomResolver | undefined, any]
+}
+
+interface IError<rethinkType, pgType> {
+  [key: string]:
+    | {
+        error: string | string[]
+        rethinkRow?: Partial<rethinkType>
+        pgRow?: Partial<pgType>
+      }
+    | number
+  recordsCompared: number
+}
+
+function getPairNeFields<rethinkType, pgType>(
+  rethinkUser: rethinkType,
+  pgUser: pgType,
+  alwaysDefinedFieldsCustomResolvers: AlwaysDefinedFieldsCustomResolvers<rethinkType>,
+  maybeUndefinedFieldsCustomResolversDefaultValues: MaybeUndefinedFieldsCustomResolversDefaultValues<
+    rethinkType
+  >
+): string[] {
+  const neFields = [] as string[]
+
+  for (const [f, customResolver] of Object.entries(alwaysDefinedFieldsCustomResolvers)) {
+    const [rethinkValue, pgValue] = [rethinkUser[f], pgUser[f]]
+    if (!lodash.isEqualWith(rethinkValue, pgValue, customResolver as CustomResolver)) {
+      neFields.push(f)
+    }
+  }
+  for (const [f, tuple] of Object.entries(maybeUndefinedFieldsCustomResolversDefaultValues)) {
+    const [customResolver, defaultValue] = tuple as [CustomResolver, any]
+    const [rethinkValue, pgValue] = [rethinkUser[f], pgUser[f]]
+    if (!lodash.isUndefined(rethinkValue)) {
+      if (!lodash.isEqualWith(rethinkValue, pgValue, customResolver)) {
+        neFields.push(f)
+      }
+    } else {
+      if (!lodash.isEqual(pgValue, defaultValue)) {
+        neFields.push(f)
+      }
+    }
+  }
+
+  return neFields
+}
+
+function addNeFieldsToErrors<rethinkType, pgType>(
+  errors: IError<rethinkType, pgType>,
+  neFields: string[],
+  rethinkRow: rethinkType,
+  pgRow: pgType,
+  rowId: string
+): void {
+  const rethinkNeFields = {}
+  const pgNeFields = {}
+  for (const neField of neFields) {
+    rethinkNeFields[neField] = rethinkRow[neField]
+    pgNeFields[neField] = pgRow[neField]
+  }
+  errors[rowId] = {
+    error: neFields,
+    rethinkRow: rethinkNeFields,
+    pgRow: pgNeFields
+  }
+}
+
+export async function checkTableEq<rethinkType, pgType>(
+  rethinkQuery: RTable<TableSchema>,
+  pgQuery: (ids: string[]) => Promise<pgType[]>,
+  alwaysDefinedFieldsCustomResolvers: AlwaysDefinedFieldsCustomResolvers<rethinkType>,
+  maybeUndefinedFieldsCustomResolversDefaultValues: MaybeUndefinedFieldsCustomResolversDefaultValues<
+    rethinkType
+  >,
+  batchSize = 3000,
+  startPage = 0
+): Promise<IError<rethinkType, pgType>> {
+  const errors = {} as IError<rethinkType, pgType>
+  const offset = batchSize * startPage
+  const rethinkRows = (await rethinkQuery
+    .skip(offset)
+    .limit(batchSize)
+    .run()) as rethinkType[]
+
+  const ids = rethinkRows.map((t) => (t as WithId<rethinkType>).id) as string[]
+  const pgRows = await pgQuery(ids)
+  const pgRowsById = {} as {[key: string]: pgType}
+  pgRows.forEach((pgRow) => {
+    pgRowsById[(pgRow as WithId<pgType>).id] = pgRow
+  })
+
+  for (const rethinkRow of rethinkRows) {
```

Comment:
implicitly, this is assuming that PG is a subset of rethinkDB.
it's possible to introduce a bug where there are more PG docs than rethinkDB docs.
as a final check, we need to get a count of all the docs in rethinkdb and all the docs in PG and make sure the 2 numbers match.

Reply:
the user `email` field also complicates this too. in pg, we set a limit of 500 chars for email field. that caused some rows to violate the unique constraint bc there were some bad actors who inserted users with ~1mb email addresses, and the first 500 chars of these were the same chars. in the backfill, i got around that problem by skipping over any users whose email had more than 500 chars after verifying that all of them looked like nonsense.

so in the case of users, the final count of both tables in rethink and pg would be different. ways around this are to add in extra logic to the fn to filter rethink rows to the valid ones. alternatively, we could drop the constraint until the migration is complete and add it before we switch over to using postgres.

more to consider on whether to defer having pg constraints @mattkrick

Reply:
added `extraPgRows`, `missingPgRows`, `rethinkRowCount`, and `pgRowCount` to the error resp. it's possible to have `rethinkRowCount === pgRowCount` and still have errors (e.g. if there's one missing pgRow and one extra pgRow), so any missing and extra rows are returned explicitly.